crates/burn-core/src/nn/rnn/lstm.rs

```diff
@@ -81,25 +64,43 @@ impl LstmConfig {
 impl<B: Backend> Lstm<B> {
     /// Applies the forward pass on the input tensor. This LSTM implementation
     /// returns the cell state and hidden state for each element in a sequence (i.e., across `seq_length`),
-    /// producing 3-dimensional tensors where the dimensions represent [batch_size, sequence_length, hidden_size].
+    /// producing 3-dimensional tensors where the dimensions represent `[batch_size, sequence_length, hidden_size]`.
     ///
-    /// Parameters:
-    ///     batched_input: The input tensor of shape [batch_size, sequence_length, input_size].
-    ///     state: An optional tuple of tensors representing the initial cell state and hidden state.
-    ///            Each state tensor has shape [batch_size, hidden_size].
+    /// ## Parameters:
+    /// - batched_input: The input tensor of shape `[batch_size, sequence_length, input_size]`.
+    /// - state: An optional tuple of tensors representing the initial cell state and hidden state.
+    ///            Each state tensor has shape `[batch_size, hidden_size]`.
     ///            If no initial state is provided, these tensors are initialized to zeros.
     ///
-    /// Returns:
-    ///     A tuple of tensors, where the first tensor represents the cell states and
-    ///     the second tensor represents the hidden states for each sequence element.
-    ///     Both output tensors have the shape [batch_size, sequence_length, hidden_size].
+    /// ## Returns:
+    /// A tuple of tensors, where the first tensor represents the cell states and
+    /// the second tensor represents the hidden states for each sequence element.
+    /// Both output tensors have the shape `[batch_size, sequence_length, hidden_size]`.
     pub fn forward(
         &self,
         batched_input: Tensor<B, 3>,
         state: Option<(Tensor<B, 2>, Tensor<B, 2>)>,
```

Comment:
Eventhough that's a breaking change, I think it might be beneficial to create a type for the state.

```rust
pub struct LstmState {
    pub cell: Tensor<B, 2>,
    pub hidden: Tensor<B, 2>,
}
```

We can remove the optional and implement `Default` instead. It also gives us a space to document what each element of the state is used for.

We could also do the same for the return type, where it's easy to make mistake (chosing the wrong returned tensor).

What do you think?

Reply:
I quite like that idea, especially for the return type!

Reply:
If breaking changes are acceptable, I'd like to follow PyTorch's return values `output, (h_n, c_n)`.

![image](https://github.com/tracel-ai/burn/assets/8537285/477a01b7-00ec-4483-9c88-860e926a6bc6)

In that case, we can return `(output, state)`. For `LSTM`, `output` and `state.hidden` are the same; for `BiLSTM`, `output` is a concatenated tensor, and the cell state and hidden state can be provided in `state` to the user without modification.

Reply:
If you agree to the modifications according to the PyTorch outputs, `batched_cell_state` is unnecessary. For the state, we only need to return the hidden state and cell state of the last time step.